apis/infrastructure/v1beta1/byohost_webhook_internal_test.go

```diff
@@ -0,0 +1,198 @@
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "host1",
+					Namespace: "default",
+				},
+				Spec: ByoHostSpec{},
+			}
+			byoHostRaw, err = json.Marshal(byoHost)
+			Expect(err).ShouldNot(HaveOccurred())
+		})
+		It("Should reject update request from invalid user", func() {
+			admissionRequest := admissionv1.AdmissionRequest{
+				Operation: admissionv1.Update,
+				UserInfo:  v1.UserInfo{Username: "unauthorized-user"},
+				Object: runtime.RawExtension{
+					Raw:    byoHostRaw,
+					Object: byoHost,
+				},
+			}
+			resp := v.Handle(ctx, admission.Request{AdmissionRequest: admissionRequest})
+			Expect(resp.AdmissionResponse.Allowed).To(Equal(false))
+			Expect(string(resp.AdmissionResponse.Result.Reason)).To(Equal(fmt.Sprintf("%s is not a valid agent username", "unauthorized-user")))
+		})
+		It("Should allow update request from manager", func() {
+			admissionRequest := admissionv1.AdmissionRequest{
+				Operation: admissionv1.Update,
+				UserInfo:  v1.UserInfo{Username: managerServiceAccount},
+				Object: runtime.RawExtension{
+					Raw:    byoHostRaw,
+					Object: byoHost,
+				},
+			}
+			resp := v.Handle(ctx, admission.Request{AdmissionRequest: admissionRequest})
+			Expect(resp.AdmissionResponse.Allowed).To(Equal(true))
+		})
+		It("Should reject request from another agent user in the group", func() {
+			admissionRequest := admissionv1.AdmissionRequest{
+				Operation: admissionv1.Update,
+				UserInfo:  v1.UserInfo{Username: "byoh:host:host2"},
+				Object: runtime.RawExtension{
+					Raw:    byoHostRaw,
+					Object: byoHost,
+				},
+			}
+			resp := v.Handle(ctx, admission.Request{AdmissionRequest: admissionRequest})
+			Expect(resp.AdmissionResponse.Allowed).To(Equal(false))
+			Expect(string(resp.AdmissionResponse.Result.Reason)).To(Equal(fmt.Sprintf("%s cannot create/update resource %s", "byoh:host:host2", "host1")))
+		})
+		It("Should allow request from the valid agent user", func() {
+			admissionRequest := admissionv1.AdmissionRequest{
+				Operation: admissionv1.Update,
+				UserInfo:  v1.UserInfo{Username: "byoh:host:host1"},
+				Object: runtime.RawExtension{
+					Raw:    byoHostRaw,
+					Object: byoHost,
+				},
+			}
+			resp := v.Handle(ctx, admission.Request{AdmissionRequest: admissionRequest})
+			Expect(resp.AdmissionResponse.Allowed).To(Equal(true))
+		})
+	})
+	Context("When ByoHost gets an delete request", func() {
+		var (
+			byoHost    *ByoHost
+			byoHostRaw []byte
+			ctx        context.Context
+		)
+		BeforeEach(func() {
+			ctx = context.TODO()
+			byoHost = &ByoHost{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ByoHost",
+					APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "host1",
+					Namespace: "default",
+				},
+				Spec: ByoHostSpec{},
+			}
+			byoHostRaw, err = json.Marshal(byoHost)
+			Expect(err).ShouldNot(HaveOccurred())
+		})
+		It("Should allow delete request from any user", func() {
```

Comment:
We should also add a test case for the following check?
```
cannot delete ByoHost when MachineRef is assigned
```

Reply:
yup good to include in the internal tests as well.

Reply:
added.

Reply:
Thanks